Running kustomize build panics

**Describe the bug**
Panic when building a project with an entry `namespace: `. `kubectl apply` doesn't complain about this, and it's hard to debug what causes it.

```
panic: interface conversion: interface {} is nil, not string

goroutine 1 [running]:
sigs.k8s.io/kustomize/api/resmap.getNamespacesForRoleBinding(0xc000355270)
        sigs.k8s.io/kustomize/api@v0.10.0/resmap/reswrangler.go:439 +0x21b
sigs.k8s.io/kustomize/api/resmap.(*resWrangler).SubsetThatCouldBeReferencedByResource(0xc00000f500, 0x4735da0)
        sigs.k8s.io/kustomize/api@v0.10.0/resmap/reswrangler.go:399 +0x105
sigs.k8s.io/kustomize/api/internal/accumulator.(*nameReferenceTransformer).Transform(0xc000ac10f0, {0x475bbd0, 0xc00000f500})
        sigs.k8s.io/kustomize/api@v0.10.0/internal/accumulator/namereferencetransformer.go:55 +0xdc
sigs.k8s.io/kustomize/api/internal/accumulator.(*ResAccumulator).Transform(...)
        sigs.k8s.io/kustomize/api@v0.10.0/internal/accumulator/resaccumulator.go:142
sigs.k8s.io/kustomize/api/internal/accumulator.(*ResAccumulator).FixBackReferences(0xd0)
        sigs.k8s.io/kustomize/api@v0.10.0/internal/accumulator/resaccumulator.go:168 +0xc5
sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).makeCustomizedResMap(0xc0003860d0)
        sigs.k8s.io/kustomize/api@v0.10.0/internal/target/kusttarget.go:131 +0x6f
sigs.k8s.io/kustomize/api/internal/target.(*KustTarget).MakeCustomizedResMap(...)
        sigs.k8s.io/kustomize/api@v0.10.0/internal/target/kusttarget.go:112
sigs.k8s.io/kustomize/api/krusty.(*Kustomizer).Run(0xc002327d98, {0x4755d60, 0x4b84ac8}, {0x7ffeefbff1b0, 0x1d})
        sigs.k8s.io/kustomize/api@v0.10.0/krusty/kustomizer.go:88 +0x39c
sigs.k8s.io/kustomize/kustomize/v4/commands/build.NewCmdBuild.func1(0xc000356580, {0xc000317900, 0x1, 0x1})
        sigs.k8s.io/kustomize/kustomize/v4/commands/build/build.go:80 +0x176
github.com/spf13/cobra.(*Command).execute(0xc000356580, {0xc0003178d0, 0x1, 0x1})
        github.com/spf13/cobra@v1.0.0/command.go:842 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0xc000356000)
        github.com/spf13/cobra@v1.0.0/command.go:950 +0x3ad
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
        sigs.k8s.io/kustomize/kustomize/v4/main.go:14 +0x1e
```

**Resources**
```
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: test
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: test
rules: []
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: test
subjects:
- kind: ServiceAccount
  name: test
  namespace:
roleRef:
  kind: Role
  name: test
  apiGroup: rbac.authorization.k8s.io
```

**Kustomize version**

```
{Version:kustomize/v4.4.0 GitCommit:63ec6bdb3d737a7c66901828c5743656c49b60e1 BuildDate:2021-09-27T16:13:36Z GoOs:darwin GoArch:amd64}
```

on macOS.

--- a/api/internal/accumulator/namereferencetransformer.go
+++ b/api/internal/accumulator/namereferencetransformer.go
@@ -52,7 +52,10 @@ func (t *nameReferenceTransformer) Transform(m resmap.ResMap) error {
 	fMap := t.determineFilters(m.Resources())
 	debug(fMap)
 	for r, fList := range fMap {
-		c := m.SubsetThatCouldBeReferencedByResource(r)
+		c, err := m.SubsetThatCouldBeReferencedByResource(r)
+		if err != nil {
+			return err
+		}
 		for _, f := range fList {
 			f.Referrer = r
 			f.ReferralCandidates = c

--- a/api/krusty/blankvalues_test.go
+++ b/api/krusty/blankvalues_test.go
@@ -1,0 +1,49 @@
+package krusty_test
+
+import (
+	"strings"
+	"testing"
+
+	kusttest_test "sigs.k8s.io/kustomize/api/testutils/kusttest"
+)
+
+// test for https://github.com/kubernetes-sigs/kustomize/issues/4240
+func TestBlankNamespace4240(t *testing.T) {
+	th := kusttest_test.MakeHarness(t)
+	th.WriteK(".", `
+resources:
+- resource.yaml
+`)
+
+	th.WriteF("resource.yaml", `
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: test
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: test
+rules: []
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: test
+subjects:
+- kind: ServiceAccount
+  name: test
+  namespace:
+roleRef:
+  kind: Role
+  name: test
+  apiGroup: rbac.authorization.k8s.io	
+`)
+
+	err := th.RunWithErr(".", th.MakeDefaultOptions())
+	if !strings.Contains(err.Error(), "Invalid Input: namespace is blank for resource") {
+		t.Fatalf("unexpected error: %q", err)
+	}
+}

--- a/api/resmap/resmap.go
+++ b/api/resmap/resmap.go
@@ -212,7 +212,7 @@ type ResMap interface {
 	// This is a filter; it excludes things that cannot be
 	// referenced by the resource, e.g. objects in other
 	// namespaces. Cluster wide objects are never excluded.
-	SubsetThatCouldBeReferencedByResource(*resource.Resource) ResMap
+	SubsetThatCouldBeReferencedByResource(*resource.Resource) (ResMap, error)
 
 	// DeAnchor replaces YAML aliases with structured data copied from anchors.
 	// This cannot be undone; if desired, call DeepCopy first.

--- a/api/resmap/reswrangler.go
+++ b/api/resmap/reswrangler.go
@@ -436,7 +436,9 @@ func getNamespacesForRoleBinding(r *resource.Resource) map[string]bool {
 		if ns, ok1 := subject["namespace"]; ok1 {
 			if kind, ok2 := subject["kind"]; ok2 {
 				if kind.(string) == "ServiceAccount" {
-					result[ns.(string)] = true
+					if n, ok3 := ns.(string); ok3 {
+						result[n] = true
+					}
 				}
 			}
 		}

--- a/api/resmap/reswrangler.go
+++ b/api/resmap/reswrangler.go
@@ -389,14 +389,17 @@ func (m *resWrangler) makeCopy(copier resCopier) ResMap {
 
 // SubsetThatCouldBeReferencedByResource implements ResMap.
 func (m *resWrangler) SubsetThatCouldBeReferencedByResource(
-	referrer *resource.Resource) ResMap {
+	referrer *resource.Resource) (ResMap, error) {
 	referrerId := referrer.CurId()
 	if referrerId.IsClusterScoped() {
 		// A cluster scoped resource can refer to anything.
-		return m
+		return m, nil
 	}
 	result := newOne()
-	roleBindingNamespaces := getNamespacesForRoleBinding(referrer)
+	roleBindingNamespaces, err := getNamespacesForRoleBinding(referrer)
+	if err != nil {
+		return nil, err
+	}
 	for _, possibleTarget := range m.rList {
 		id := possibleTarget.CurId()
 		if id.IsClusterScoped() {
@@ -416,20 +419,20 @@ func (m *resWrangler) SubsetThatCouldBeReferencedByResource(
 			result.append(possibleTarget)
 		}
 	}
-	return result
+	return result, nil
 }
 
 // getNamespacesForRoleBinding returns referenced ServiceAccount namespaces
 // if the resource is a RoleBinding
-func getNamespacesForRoleBinding(r *resource.Resource) map[string]bool {
+func getNamespacesForRoleBinding(r *resource.Resource) (map[string]bool, error) {
 	result := make(map[string]bool)
 	if r.GetKind() != "RoleBinding" {
-		return result
+		return result, nil
 	}
 	//nolint staticcheck
 	subjects, err := r.GetSlice("subjects")
 	if err != nil || subjects == nil {
-		return result
+		return result, nil
 	}
 	for _, s := range subjects {
 		subject := s.(map[string]interface{})
@@ -438,12 +441,14 @@ func getNamespacesForRoleBinding(r *resource.Resource) map[string]bool {
 				if kind.(string) == "ServiceAccount" {
 					if n, ok3 := ns.(string); ok3 {
 						result[n] = true
+					} else {
+						return nil, errors.New("Invalid Input: namespace is blank")
 					}
 				}
 			}
 		}
 	}
-	return result
+	return result, nil
 }
 
 // AppendAll implements ResMap.

--- a/api/resmap/reswrangler.go
+++ b/api/resmap/reswrangler.go
@@ -442,7 +442,7 @@ func getNamespacesForRoleBinding(r *resource.Resource) (map[string]bool, error) 
 					if n, ok3 := ns.(string); ok3 {
 						result[n] = true
 					} else {
-						return nil, errors.New("Invalid Input: namespace is blank")
+						return nil, errors.New(fmt.Sprintf("Invalid Input: namespace is blank for resource %q\n", r.CurId()))
 					}
 				}
 			}

--- a/api/resmap/reswrangler_test.go
+++ b/api/resmap/reswrangler_test.go
@@ -566,7 +566,10 @@ func TestSubsetThatCouldBeReferencedByResource(t *testing.T) {
 	for name, test := range tests {
 		test := test
 		t.Run(name, func(t *testing.T) {
-			got := m.SubsetThatCouldBeReferencedByResource(test.filter)
+			got, err1 := m.SubsetThatCouldBeReferencedByResource(test.filter)
+			if err1 != nil {
+				t.Fatalf("Expected error %v: ", err1)
+			}
 			err := test.expected.ErrorIfNotEqualLists(got)
 			if err != nil {
 				test.expected.Debug("expected")

